### PR TITLE
[JIT] linear scan memory planning strategy

### PR DESCRIPTION
**Pull request by makslevental**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63873
* #64350
* #64349
* __->__ #64348
* #64347
* #62995

Linear scan (a heuristic based on http://web.cs.ucla.edu/~palsberg/course/cs132/linearscan.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

### CODE REVIEW

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769099).

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      // split it and keep that available space for other live ranges
+      if (candidate_size - aligned_curr_size > 0) {
+        avail_regions.insert(MemRegion{
+            candidate_offset + aligned_curr_size,
+            candidate_size - aligned_curr_size});
+      }
+      continue;
+    }
+
+    TORCH_INTERNAL_ASSERT(
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/memory_planning/linear_scan.cpp:94]: lmao nuclear shelter proof 😸

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+              item.second.size >= aligned_curr_size;
+        });
+
+    if (swap_candidate != active.rend()) {
+      auto candidate_reg = swap_candidate->second;
+      size_t candidate_offset = candidate_reg.offset;
+      size_t candidate_size = candidate_reg.size;
+      active[curr_range] = {candidate_offset, aligned_curr_size};
+      // split region (potentially)
+      if (candidate_size > aligned_curr_size) {
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/memory_planning/linear_scan.cpp:113]: looks like duplication of the logic above? Maybe we should refactor it into a helper?

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    TORCH_INTERNAL_ASSERT(
+        candidate_reg == avail_regions.end() && swap_candidate == active.rend(),
+        "inconsistent state");
+
+    // create new alloc
+    active[curr_range] = {curr_end_offset, aligned_curr_size};
+    curr_end_offset += aligned_curr_size;
+  }
+
+  // expire any remaining intervals
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/memory_planning/linear_scan.cpp:134]: I feel like this comment is very helpful, but I struggle to come up with something better. The only side effect you care about here is moving remaining intervals to `allocated_ranges` .

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  allocate_inactive_ranges(
+      {std::numeric_limits<size_t>::max(), std::numeric_limits<size_t>::max()});
+
+  std::vector<MemAllocation> allocations;
+  allocations.reserve(allocated_ranges.size());
+  for (const auto& item : allocated_ranges) {
+    allocations.push_back({item.first, item.second});
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/memory_planning/linear_scan.cpp:141]: err this looks fishy... you already reserved enough elements for `allocated_ranges` but then you are pushing back those on top reserved ones?

File: torch/csrc/jit/passes/memory_planning/linear_scan.h [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <torch/csrc/jit/passes/memory_planning.h>
+#include <torch/csrc/jit/runtime/static/impl.h>
+
+namespace torch {
+namespace jit {
+
+using EndSortedLiveRangeMap =
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/memory_planning/linear_scan.h:9]: it's only used in the `cpp` file, why don't we move it there?

File: test/cpp/jit/test_memory_planning.cpp [hunk 6]
```diff
@@ -432,5 +432,32 @@ TEST(MemoryPlannerTest, LSTMNaive) {
...
 }
 
+TEST(MemoryPlannerTest, SmallLinearScan) {
+  StorageAttrs expected_storage = {1344, DeviceType::CPU};
+  std::vector<AllocAttrs> expected_allocs = {
+      {448, 0, TTP(((Vec{10, 10})), ((Vec{10, 1})))},
+      {448, 448, TTP(((Vec{10, 10})), ((Vec{10, 1})))},
+      {448, 896, TTP(((Vec{10, 10})), ((Vec{10, 1})))},
+  };
+  testSmall(expected_storage, expected_allocs, Strategy::LINEAR_SCAN);
```
Comment (Krovatkin) [inline test/cpp/jit/test_memory_planning.cpp:442]: I take it all strategies are tested with only these two?

**Review by Krovatkin — APPROVED** [review]
:shipit: Could you please take a look at https://github.com/pytorch/pytorch/pull/64348#discussion_r714344127 before landing this?

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+// aren't fixed "width" like registers
+std::vector<MemAllocation> linearScanHeuristic(
+    SortedLiveRangeMap<size_t> live_ranges) {
+  // sorted by right endpoint so we can find latest ending range quickly
+  EndSortedLiveRangeMap active;
+  std::multiset<MemRegion, regionSizeCmp> avail_regions;
+  SortedLiveRangeMap<MemRegion> allocated_ranges;
+
+  size_t curr_end_offset = 0;
+  auto allocate_inactive_ranges = [&](UniqueLiveRange curr_range) {
```
[the remark this answers is not in the document]
Comment (makslevental) [reply torch/csrc/jit/passes/memory_planning/linear_scan.cpp:37]: agreed. will do

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    auto curr_size = item.second;
+    allocate_inactive_ranges(curr_range);
+    coalesce_avail(avail_regions);
+
+    auto aligned_curr_size = MemoryPlanner::compute_aligned_tensor_size(curr_size);
+
+    // find the "right" region; in order of preference:
+    // 1. tightest fit free region i.e. smallest i.e. first match since
+    // avail_regions is sorted by size
+    // 2. swap with latest ending active live range that is big enough (spilling
```
[the remark this answers is not in the document]
Comment (makslevental) [reply torch/csrc/jit/passes/memory_planning/linear_scan.cpp:64]: i mean it probably doesn't if the data is indicative but this is the heuristic from the original linear scan paper


Screenshot from 2021-09-22 17-20-11

File: test/cpp/jit/test_memory_planning.cpp (hunk 6 above)
Comment (makslevental) [reply test/cpp/jit/test_memory_planning.cpp:442]: correct. longer models would've inflated the verification spec data too much i think

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp (hunk 1 above)
Comment (makslevental) [reply torch/csrc/jit/passes/memory_planning/linear_scan.cpp:94]: is this the wrong thing? should i use something else?

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp (hunk 2 above)
Comment (makslevental) [reply torch/csrc/jit/passes/memory_planning/linear_scan.cpp:113]: it's true but it's mostly duplication of unpacking structs and such. only duplicated logic is the check which is more intelligible in context i think. not sure - could go either way (we should flip a coin)

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp (hunk 3 above)
Comment (makslevental) [reply torch/csrc/jit/passes/memory_planning/linear_scan.cpp:134]: discussed offline

Comment (makslevental) [reply torch/csrc/jit/passes/memory_planning/linear_scan.cpp:134]: i think you meant *not very helpful? agreed. will add a couple more words

File: torch/csrc/jit/passes/memory_planning/linear_scan.cpp (hunk 4 above)
Comment (makslevental) [reply torch/csrc/jit/passes/memory_planning/linear_scan.cpp:141]: discussed offline

File: torch/csrc/jit/passes/memory_planning/linear_scan.h (hunk 5 above)
Comment (makslevental) [reply torch/csrc/jit/passes/memory_planning/linear_scan.h:9]: i can't remember if it's used in any of the other diffs but it's a vaguely useful typedef? not sure. another coin flip.

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769099).

**Pull request closed without merging**